sdk/storage/azure-storage-quickquery/src/main/java/com/azure/storage/quickquery/implementation/util/FluxInputStream.java

```diff
@@ -0,0 +1,232 @@
+        // Block current thread until data is available.
+        while (waitingForData) {
+            if (fluxComplete) {
+                break;
+            } else {
+                try {
+                    dataAvailable.await();
+                } catch (InterruptedException e) {
+                    throw logger.logExceptionAsError(new RuntimeException(e));
+                }
+            }
+        }
+        lock.unlock();
+    }
+
+    /**
+     * Subscribes to the data with a special subscriber.
+     */
+    private void subscribeToData() {
+        this.data
+            .onBackpressureBuffer(Constants.GB)
+            .subscribe(
+            // ByteBuffer consumer
+            byteBuffer -> {
+                lock.lock();
+                this.buffer = new ByteArrayInputStream(FluxUtil.byteBufferToArray(byteBuffer));
+                this.waitingForData = false;
+                // Signal the consumer when data is available.
+                dataAvailable.signal();
+                lock.unlock();
+            },
+            // Error consumer
+            throwable -> {
+                // Signal the consumer in case an error occurs (indicates we completed without data).
+                signalOnCompleteOrError();
+                if (throwable instanceof BlobStorageException) {
+                    this.lastError = new IOException(throwable);
+                } else if (throwable instanceof IllegalArgumentException) {
+                    this.lastError = new IOException(throwable);
+                }
+            },
+            // Complete consumer
+                // Signal the consumer in case we completed without data.
+                this::signalOnCompleteOrError,
+            // Subscription consumer
+            subscription -> {
+                this.subscription = subscription;
+                this.subscribed = true;
+                this.subscription.request(1);
+            }
+        );
+    }
+
+    /**
+     * Signals to the subscriber when the flux completes without data (onCompletion or onError)
+     */
+    private void signalOnCompleteOrError() {
+        lock.lock();
+        this.fluxComplete = true;
+        this.waitingForData = false;
+        dataAvailable.signal();
+        lock.unlock();
+    }
+
+    /**
+     * Validates parameters according to {@link InputStream#read(byte[], int, int)} spec.
+     *
+     * @param b the buffer into which the data is read.
+     * @param off the start offset in array b at which the data is written.
+     * @param len the maximum number of bytes to read.
+     */
+    private void validateParameters(byte[] b, int off, int len) {
+        if (b == null) {
+            throw new NullPointerException("'b' cannot be null");
+        }
+        if (off < -1) {
+            throw new IndexOutOfBoundsException("'off' cannot be less than -1");
+        }
+        if (len < -1) {
+            throw new IndexOutOfBoundsException("'len' cannot be less than -1");
+        }
```

Comment:
Shouldn't these be checking `< 0` and the exception stating that these cannot be `less than 0`?

Reply:
you're right this should be less than or equal to. My bad

Reply:
I'll be eventually putting out another PR to put this in the blobs package. I'll fix it then.